### ISSUE TITLE
Tracking Issue for denying trailing semicolons in expression macro bodies

### ISSUE THREAD

**Issue by Aaron1011** [rust-lang/rust#79813]
This is a tracking issue for MCP https://github.com/rust-lang/lang-team/issues/70

### What is this issue?

If you're a crate author who's been linked here, this issue tracks emitting a hard error when a trailing semicolon occurs in a macro used in expression position. See https://github.com/rust-lang/lang-team/issues/70 for some background about the issue.

In the future, code like this may stop compiling:

```rust
macro_rules! foo {
    () => {
        true;
    }
}

fn main() {
    let val = match true {
        true => false,
        _ => foo!()
    };
}
```

In the above code snippet, `foo!()` is invoked in expression position (in this case, a macro arm). However, `foo!()` expands to the token stream `true;`, which is not a valid expression due to the presence of a trailing semicolon.

To make the behavior of macro expansion more consistent, we plan to turn this into a hard error at some point in the future.

Fixing issues in your crate is usually as simple as removing the trailing semicolon from the macro definition.

Currently, the associated lint (`semicolon_in_expressions_from_macros`) is set to allow-by-default. As time goes on, we may increase this to warn-by-default.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] Implement the lint: https://github.com/rust-lang/rust/pull/79819
- [X] Do a Crater run to determine the full extent of the impact (https://github.com/rust-lang/rust/pull/81477)
- [X] Fix `lint_node_id` so that we can apply `#[allow]` close to a macro call site (https://github.com/rust-lang/rust/pull/87146)
- [ ] When https://github.com/rust-lang/rustfmt/pull/4507 makes it way into a stable release, make the lint warn-by-default
- [ ] Mark the lint for inclusion in the cargo future-incompat-report (see https://github.com/rust-lang/rust/issues/71249)
- [ ] Switch the lint to deny-by-default
- [ ] Consider making the lint a hard error over an edition boundary

### Implementation history

1. Lint implemented in https://github.com/rust-lang/rust/pull/79819

**Comment by nikomatsakis** [comment]
Hey @Aaron1011! Sorry for the late notice, but any chance you can post an update as a comment on this issue sometime before the lang-team planning meeting tomorrow? [Here are some prompts and questions.](https://hackmd.io/BurZQXKST6i0vFVWn61Vaw?view#Questions-to-consider-for-updates)

**Comment by Aaron1011** [comment]
Sure!

This is currently blocked on bumping the `stdarch` submodule to include https://github.com/rust-lang/stdarch/pull/1080. Once that's done, I'll do a Crater run to determine how many crates need to be updated.

Before we make this warn-by-default, we'll probably want to fix `lint_node_id` to allow placing `semicolon_in_expressions_from_macros` closer to the offending macro call - currently, it can only be placed on the enclosing item definition (e.g. a function).

Design-wise, I don't think there's anything that needs to be done - this lint is about removing a special check for the macro parsing code.

**Comment by nikomatsakis** [comment]
@Aaron1011 any updates for the planning meeting this month?

**Comment by Aaron1011** [comment]
@nikomatsakis: https://github.com/rust-lang/rust/pull/83089 was recently merged, so we're now denying the lint within the compiler and standard library. The next step is to do a Crater run, which I'll kick off.

**Comment by Aaron1011** [comment]
The crater run https://github.com/rust-lang/rust/pull/81477 showed 25k regressions, with over 1k root regressions. 
I've opened https://github.com/rust-lang/rust/pull/87146 to improve `lint_node_id`, which should let us make this warn-by-default at some point.

**Comment by Fishrock123** [comment]
I think this is probably unnecessary churn in situations where the semicolon follows anything which returns `!`, e.g. `return`, `continue`, `break`, `panic!` or the like.

I think the compiler should be able to detect this and ignore it since if that exists the output of the macro will always be `!` regardless of the semicolon.

This was noticed in http-types:
```
warning: trailing semicolon in macro used in expression position
   --> src/macros.rs:5:63
    |
5   |         return $crate::private::Err($crate::format_err!($msg));
    |                                                               ^
    |
   ::: src/utils/date.rs:100:18
    |
100 |             _ => bail!("Invalid Month"),
    |                  ---------------------- in this macro invocation
    |
    = note: `#[warn(semicolon_in_expressions_from_macros)]` on by default
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #79813 <https://github.com/rust-lang/rust/issues/79813>
    = note: this warning originates in the macro `bail` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Commit-pinned code: https://github.com/http-rs/http-types/blob/25d6af1edd82f25791d5795cc815929ac64f346a/src/macros.rs#L1-L13

We certainly are not the only ones to have such macros, other libraries like `anyhow` also have return-style macros such as `anyhow::bail!`.

**Comment by Aaron1011** [comment]
@Fishrock123: The idea behind this change is to make expansion more 'token-based' - the result of a macro call should (with some exceptions) be the same as copy-pasting the macro body, and replacing all of the metavariables by hand.

In your example:

```rust
_ => bail!("Invalid Month"),
```

will expand to

```rust
_ => return $crate::private::Err($crate::format_err!("Invalid Month"));

**Comment by devsnek** [comment]
I have some macros like this:

```rust
macro_rules! something {
  ($macro:ident) => {
     macro!(
       a
       lot
       of stuff
       i don't want to copy between different macro invocations
     );
  }
}
```

The issue is that sometimes `macro` expands to an item and sometimes it expands to an expression, so either i have a semicolon at the end and get `trailing semicolon in macro used in expression position` or i remove the semicolon and get `macros that expand to items must be delimited with braces or followed by a semicolon`.

How this this code be migrated to deal with `semicolon_in_expressions_from_macros` becoming a hard error in the future?

**Comment by Aaron1011** [comment]
@joshtriplett: I've been letting this slowly percolate through the ecosystem (there's a steady stream of commits referencing this issue), as macro-related lints are always tricky for users to deal with. However, I think enough time has passed that we could add this to the future-incompat report.

**Comment by vitalytarasov** [comment]
i was a little perplex landing on this page from rustc warning for a macro defined as following:
```
macro_rules! debug {
    ($fmt:expr) => {
        debug!("{}", &format!($fmt))
    };
    ($fmt:expr, $($arg:tt)+) => {
        crate::logger::log(crate::logger::Level::Debug, &format!($fmt, $($arg)+))
    }
}
```

how does this results in a trailing semicolon?

**Comment by John-Nagle** [comment]
Breaks some older versions of "nom", which breaks an old version of "iso8601" (date parsing),  which breaks "quick-xml", which breaks "simple-xmlrpc". Seems to be fixed in Github but new version of "simple-xmlrpc"  has not been pushed to "crates.io". Probably because it was complete, stable, had no issues, and had not needed maintenance in four years, so no one is maintaining it.

Ref: https://github.com/belak/serde-xmlrpc/issues

**Comment by sffc** [comment]
I have code where it's not evident to me how to write it in a way that avoids this warning. It is a macro that takes another macro as a callback and conditionally generates either expressions or items.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=75617c4dc910f1e7a1f044dbee8e6c9c

If I remove the semicolon, I can't generate items. If I add the semicolon, it works but I get this warning.

See https://github.com/unicode-org/icu4x/issues/4910

**Comment by dtolnay** [comment]
Macros that expand to expressions sometimes and items sometimes need to use brace syntax and no semicolon. `cb! {…}`

The error message already suggests this as a solution too.

```console
error: macros that expand to items must be delimited with braces or followed by a semicolon
  --> src/main.rs:9:13
   |
9  |           $cb!([
   |  _____________^
```

**Comment by traviscross** [comment]
We talked about this in the lang call today, and we have appetite for a PR that would move this FCW to deny-by-default (and continue to warn in dependencies).

**Comment by Thomasdezeeuw** [comment]
We've hit this warning (hard error in nightly) in Mio. However, the fix was the opposite of what this tracking issue documents in that we had to add a semicolon. See https://github.com/tokio-rs/mio/pull/1975 for details. I think it would be worth documenting this case in the main issue for people hitting it as well, ideally before this hits stable.

Edit: created https://github.com/rust-lang/rust/issues/159388.

**Comment by mati865** [comment]
@Thomasdezeeuw this is worth opening a new issue.

**Comment by VorpalBlade** [comment]
In 1.99.0-nightly this lint is pretty bad, since it triggers at deny level in downstream crates, even locally where I *don't* use `-D warnings` (unlike in CI):

```
error: trailing semicolon in macro used in expression position
   --> crates/konfigkoll_hwinfo/src/pci/parser.rs:225:37
    |
225 |             | Line::Subsystem(_) => eyre::bail!("Unexpected line at top level: {line:?}"),
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #79813 <https://github.com/rust-lang/rust/issues/79813>
    = note: `#[deny(semicolon_in_expressions_from_macros)]` (part of `#[deny(future_incompatible)]`) on by default
    = note: this error originates in the macro `eyre::bail` (in Nightly builds, run with -Z macro-backtrace for more info)
```

This means anyone just building my software with `cargo run` will get affected once this reaches stable. I don't know if it also affects `cargo install`, that would be really bad.

Perhaps the hard error needs to be rolled back until the ecosystem has time to adapt. I can't really fix upstream here, and `eyre` is really slow to release updates.

**Comment by Thomasdezeeuw** [comment]
Is there any action being taken here to reduce this to a warning? Changing from an allowed lint, i.e. no output at all, straight to an error, i.e. fail to compile, without a period of time that it's an warning (but still compiles) is quite painful. Mio for example currently fails to compile on nightly because of this. This is very much a breaking change without any grace period (for nightly users).

**Comment by mati865** [comment]
It's pending a discussion AFAICT: https://github.com/rust-lang/rust/pull/159218#issuecomment-5035452966

**Comment by joshtriplett** [comment]
Latest nightly splits the warnings on non-local macros into a separate lint `semicolon_in_expressions_from_non_local_macros`, which starts out as `warn` rather than `deny`.

If you encounter a lint like this coming from a macro not in your crate, in some cases you can work around it by invoking the macro in a different (non-expression) context, or invoking the macro using braces (`the_macro! { .. }`) rather than parentheses. In other cases, you may not be able to fix it directly, and it will need fixing in the upstream crate, so you'll likely need to `allow` it temporarily. However, before you add an `allow`, please make sure to report it to the crate providing the macro so they can fix it. (Check for duplicate reports before doing so, and for newer versions of the crate, as other users of the crate may have already reported it.)
